File: my-folder/0199-binary-tree-right-side-view/solution.java
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
      List<Integer> result = new ArrayList<>();
      
      dfs(root, 0, result);  
      
      return result;
    }
    
    private void dfs(TreeNode node, int level, List<Integer> result) {
      if (node == null) {
        return;
      }
      
      if (level >= result.size()) {
        result.add(node.val);
      } else {
        result.set(level, node.val);
      }
      
      dfs(node.left, level + 1, result);
      dfs(node.right, level + 1, result);
    }
}

File: my-folder/0199-binary-tree-right-side-view/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun rightSideView(root: TreeNode?): List<Int> {
        if (root == null) return emptyList()
        
        val list = ArrayList<Int>()
        val q = LinkedList<TreeNode>()
        q.offer(root)
        
        while(q.isNotEmpty()) {
          val qSize = q.size
          for (i in 0 until qSize) {
            val polled = q.poll()
                      
            polled.left?.let { q.offer(it) }
            polled.right?.let { q.offer(it) }
            
            if (i == qSize-1) { list.add(polled.`val`) }
          }
        }
        
        return list
    }
}
